esp-wifi/src/lib.rs

```diff
@@ -338,11 +362,138 @@ pub fn initialize(
 
     match init_for {
         #[cfg(feature = "wifi")]
-        EspWifiInitFor::Wifi => Ok(EspWifiInitialization::Wifi(EspWifiInitializationInternal)),
+        EspWifiFor::Wifi => Ok(EspWifiInitialization::Wifi(EspWifiInitializationInternal)),
+        #[cfg(feature = "ble")]
+        EspWifiFor::Ble => Ok(EspWifiInitialization::Ble(EspWifiInitializationInternal)),
+        #[cfg(coex)]
+        EspWifiFor::WifiBle => Ok(EspWifiInitialization::WifiBle(
+            EspWifiInitializationInternal,
+        )),
+    }
+}
+
+#[cfg(feature = "wifi")]
+pub fn deinitialize_wifi(
+    controller: WifiController,
+    stack: crate::wifi_interface::WifiStack<impl WifiDeviceMode>,
+) -> Result<EspWifiDeinitialization, WifiError> {
+    // Stop and deinitialize the WiFi stack
+    esp_wifi_result!(unsafe { esp_wifi_stop() })?;
+    esp_wifi_result!(unsafe { esp_wifi_deinit_internal() })?;
+
+    info!("WiFi deinitialized");
+
+    // Drop the WiFi controller and stack
+    drop(controller);
+    drop(stack);
+
+    // Return WiFi deinitialization state
+    Ok(EspWifiDeinitialization::Wifi(
+        EspWifiDeinitializationInternal,
+    ))
+}
+
+#[cfg(feature = "ble")]
+pub fn deinitialize_ble(ble: bleps::Ble) -> Result<EspWifiDeinitialization, ()> {
+    // Deinitialize BLE based on the chip type
+    #[cfg(any(esp32, esp32c3, esp32s3))]
+    crate::ble::btdm::ble_deinit();
+
+    #[cfg(any(esp32c2, esp32c6, esp32h2))]
+    crate::ble::npl::ble_deinit();
+
+    info!("BLE deinitialized");
+
+    drop(ble);
+
+    // Return BLE deinitialization state
+    Ok(EspWifiDeinitialization::Ble(
+        EspWifiDeinitializationInternal,
+    ))
+}
+
+#[cfg(coex)]
+pub fn deinitialize_all(
+    controller: WifiController,
+    stack: crate::wifi_interface::WifiStack<impl WifiDeviceMode>,
+    ble: bleps::Ble,
+) -> Result<EspWifiDeinitialization, WifiError> {
+    // Disable coexistence
+    unsafe { crate::wifi::os_adapter::coex_disable() };
+    unsafe { crate::wifi::os_adapter::coex_deinit() };
+
+    // Deinitialize WiFi
+    esp_wifi_result!(unsafe { esp_wifi_stop() })?;
+    esp_wifi_result!(unsafe { esp_wifi_deinit_internal() })?;
+
+    // Deinitialize BLE
```

Comment:
We should resuse the `deinitialize_ble` etc methods here in the deinit all

Reply:
Ideally, we shouldn't care about what network or Bluetooth stack the user is using - e.g. they can use embassy-net instead and they can use Trouble instead of bleps. Users can even directly deal with HCI and use raw `smoltcp`